docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+# CIP: Pipelines, The Registry and Storage
```

Comment:
I'm still mixed on the term pipeline ... alternatives include
- workflow (temporal)
- flow and tasks (prefect)
- transforms
- lambdas
- functions
- jobs & ops (dagster)
- chain (langchain)

Reply:
Workflow - fine, but I think its less descriptive than pipeline and evokes a different set of tools.
Flow - Refers to a generalization of a pipeline, I am fine with this. A pipeline is 1D flow.(https://en.wikipedia.org/wiki/Pipeline_(software))
Transforms - not all encompassing, some of these have side effects or are not transformations
Lambdas - feels really specific to something not explicitly always used here, especially in python, and does not communicate sequentiality. 
Functions - doesn't communicate sequence.
Jobs & Ops - Dagsters model is far more complicated (Assets, graphs etc). I think a "Job" is a bit broad and an "Op" is fine but seems like needlessly using language no one uses. Job feels really broad
Chain - feels owned by langchain and also is not descriptive for parallel execution. 

I prefer pipeline or flow because they are most descriptive of a sequence of operations on data, where inputs and outputs flow into one another, and potentially with side effects. Pipeline seems like something people will be familiar with, flow will be a new term.

Reply:
appreciate the reactions!

Reply:
@HammadB @jeffchuber, I too think that pipeline conveys the meaning of chaining a bunch of discrete functions together. That said why not just "invent" a name for this? We can still call it a pipeline just so people draw a parallel but then the actual "catchy" name can be something of the sort of ChromaticFunnel :D

Naming aside, I am wondering about two additional concepts that come to mind with this - filters and error syncs.
- Filters - useful when dropping out samples is a thing
- Error sinks - basically a way to "nicely" handle inevitable errors (again user-driven functions)

Reply:
i am personally generally against making up new names for things that already have good names :)

Reply:
I understand filters and thats a good point, you can imagine input filters that can detect "bad" data and filter it out.

What are error sinks? Can you elaborate?

Reply:
it is a try/catch for pipelines. 

I think @jeffchuber will be able to relate here with JS' promises:

`do_something().do_another_thing().catch(e-> do_something_with_error(e))`

Reply:
oh and another "nice" concept from Java world - parallel() - the ability to split/join a sequence of pipeline steps. This can be done further down the road but I can see how a batch of 100k docs can have parallel embedding going for it.

Reply:
ah yeah, supporting something akin to promise like error handling is a good idea

Reply:
error handling is def interesting and important

Reply:
my initial thought here was that pipelines are synchronous with the insert call, so error states are trivially propagated. Once we allow for async pipelines though this becomes much harder.

Reply:
@HammadB, I think that the error handling can be done optional e.g. if people want to do something with their failing batches - e.g. retry, dump them to disk so they can retry later etc.